_overviews/scala3-book/scala-for-javascript-devs.md

```diff
@@ -0,0 +1,1238 @@
+
+// map, flatMap
+val fruits = List("apple", "pear")
+fruits.map(_.toUpperCase)             // List(APPLE, PEAR)
+fruits.flatMap(_.toUpperCase)         // List(A, P, P, L, E, P, E, A, R)
+
+val nums = List(10, 5, 8, 1, 7)
+nums.sorted                           // List(1, 5, 7, 8, 10)
+nums.sortWith(_ < _)                  // List(1, 5, 7, 8, 10)
+nums.sortWith(_ > _)                  // List(10, 8, 7, 5, 1)
+
+List(1,2,3).updated(0,10)             // List(10, 2, 3)
+List(2,4).union(List(1,3))            // List(2, 4, 1, 3)
+
+// zip
+val women = List("Wilma", "Betty")    // List(Wilma, Betty)
+val men = List("Fred", "Barney")      // List(Fred, Barney)
+val couples = women.zip(men)          // List((Wilma,Fred), (Betty,Barney))
+```
+
+Scala has *many* more methods that are available to you. The benefits of all these methods are:
+
+- You don’t have to write custom `for` loops to solve problems
+- When you read someone else’s code, you won’t have to read their custom `for` loops; you’ll just find common methods like these, so it’s easier to read code from different projects  
+
+
+### Tuples
+
+When you want to put multiple data types in the same list, JavaScript lets you do this:
+
+```scala
+let stuff = ["Joe", 42, 1.0]
+```
+
+In Scala you do this:  
+
+```scala
+val a = ("eleven")
+val b = ("eleven", 11)
+val c = ("eleven", 11, 11.0)
+val d = ("eleven", 11, 11.0, Person("Eleven"))
+```
+
+In Scala these types are called tuples, and as shown, they can contain one or more elements, and the elements can have different types. You access their elements just like you access elements of a `List`, `Vector`, or `Array`:  
+
+```scala
+d(0)   // "eleven"
+d(1)   // 11
+```
+
+### Enumerations
+
+JavaScript doesn’t have enumerations, but you can do this:
+
+```javascript
+let Color = {
+  RED: 1,
+  GREEN: 2,
+  BLUE: 3
+};
+Object.freeze(Color);
+```
+
+In Scala 3 you can do quite a few things with enumerations. You can create an equivalent of that code:
+
+```scala
+enum Color:
+  case Red, Green, Blue
+```
+
+You can create a parameterized enum:  
+
+```scala
+enum Color(val rgb: Int):
+  case Red   extends Color(0xFF0000)
+  case Green extends Color(0x00FF00)
+  case Blue  extends Color(0x0000FF)
+```
+
+You can also create user-defined enum members:
+
+```scala
+enum Planet(mass: Double, radius: Double):
+  case Mercury extends Planet(3.303e+23, 2.4397e6)
+  case Venus   extends Planet(4.869e+24,6.0518e6)
+  case Earth   extends Planet(5.976e+24,6.37814e6)
+  // more planets here ...
+
+  private final val G = 6.67300E-11
+  def surfaceGravity = G * mass / (radius * radius)
+  def surfaceWeight(otherMass: Double) = otherMass * surfaceGravity
+```
+
+
+
+## Scala.js DOM Code
+
+Scala.js is a project that lets you write Scala code that is compiled to JavaScript code that can then be used in the browser. The approach is similar to CoffeeScript, TypeScript, and other languages that are compiled (“transpiled”) to JavaScript.  
```

Comment:
```suggestion
Scala.js lets you write Scala code that is compiled to JavaScript code that can then be used in the browser. The approach is similar to TypeScript, ReScript, and other languages that are compiled to JavaScript.  
```
(please don't say "transpiled"; this word means nothing)

Reply:
haha :) true